app/assets/javascripts/components/plots/forms/plot_form.jsx

```diff
@@ -0,0 +1,348 @@
+          </fieldset>
+          <fieldset style={{ marginBottom: 10 }}>
+            <legend>Data</legend>
+            {this.dataRefList()}
+            {this.dataFormFields()}
+            <input type='button' value='Add' onClick={() => this.addDataRef()} />
+          </fieldset>
+          <Plot plot={plot} consignment={consignment}/>
+        </div>
+      );
+    }
+  };
+};
+
+/*
+input components subscribe to a property from the plot configuration state.
+eg. subscribePlotInputField(['layout', 'height'])(IntegerField('Height: '))
+the nested value for { layout: { height: value } } is injected into the input component
+*/
+export const subscribePlotInputField = (plotProperty = []) => (WrappedInput) => (props) => {
+  let { value, plot, consignment, updatePlot, ...passThroughProps } = props;
+
+  let currentValue = plotProperty.reduce((currVal, currProperty) => {
+    try {
+      return currVal[currProperty];
+    } catch (e) {
+      return null;
+    }
+  }, plot);
+
+  if (typeof currentValue !== 'string' && isNaN(currentValue)) {
+    currentValue = '';
+  }
+
+  const update = (value) => {
+    let updatedPlot = { ...plot };
+
+    let ref = updatedPlot;
+    const finalPropertyIndex = plotProperty.length - 1;
+    for (let i = 0; i < finalPropertyIndex; i++) {
+      ref = ref[plotProperty[i]];
+    }
+
+    ref[plotProperty[finalPropertyIndex]] = value;
+
+    updatePlot(updatedPlot);
+  };
+
+  return (
+    <WrappedInput
+      {...passThroughProps}
+      consignment={consignment}
+      value={currentValue}
+      onChange={update}
+    />
+  );
+};
+
+const consignmentKeysByType = (type) => (consignment) => {
+  return Object.keys(consignment || {}).filter(k => consignment[k] instanceof type);
+};
+
+const matrixConsignmentKeyFilter = consignmentKeysByType(Matrix);
+const vectorConsignmentKeyFilter = consignmentKeysByType(Vector);
+
+const ConsignmentKeySelector = (label, filter) => ({ consignment, ...passThroughProps }) => (
+  <Select
+    label={label}
+    options={filter(consignment)}
+    hasNull={true}
+    {...passThroughProps}
+  />
+);
+
+export const MatrixSelector = (label) => ConsignmentKeySelector(label, matrixConsignmentKeyFilter);
+export const VectorSelector = (label) => ConsignmentKeySelector(label, vectorConsignmentKeyFilter);
+export const Selector = (label, options) => (props) => (
+  <Select
+    {...props}
+    label={label}
+    options={options}
+    hasNull={false}
+  />
+);
+
+export const TextField = (label) => ({ type, ...passThroughProps }) => (
+  <InputField
+    type='text'
+    label={label}
+    {...passThroughProps}
+  />
+);
+
+const IntegerField = (label) => withIntegerFilter(TextField(label));
+
+export const CheckBox = (label) => ({ type, value, onChange, ...passThroughProps }) => (
+  <InputField
+    type='checkbox'
+    label={label}
+    checked={value}
+    onChange={() => onChange(!value)}
+    {...passThroughProps}
+  />
+);
+
+export const commonfields = [
```

Comment:
This gets exported and passed right back in - could you just use it locally and spare the import/export?

Reply:
probably, I don't know how common it will actually be. I decided it makes more sense to pass it in rather than saying it should be common across all future plot types.